### PR TITLE
don't build M and w for inactive contacts/limits

### LINKED ISSUE #637: efficiently ignore lcp constraints for phi >> 0.
test case is to speed up simulation of atlas in the terrain world.

### PR DESCRIPTION
**Pull request by RussTedrake**
Resolves #637 by building smaller LCPs.  The simulation just got a lot faster.  ;)

profile results below for before and after the change using "runAtlasDynamics" as my test case:

screen shot 2015-01-24 at 8 07 15 am

### CODE REVIEW

**Comment by RussTedrake** [comment]
one thing you can see here is that fastQP is not succeeding as often anymore (lcppath is getting called more often).  I don't have a good explanation for that one, except that I have to through out the active set when the possible contacts/limits change.

**Comment by RussTedrake** [comment]
i *think* this is ready (i could not reproduce the one failing unit test).  @kuindersma could you give me a careful code review?

File: systems/plants/TimeSteppingRigidBodyManipulator.m [hunk 1]
```diff
@@ -248,6 +236,7 @@ function checkDirty(obj)
         num_q = obj.manip.num_positions;
         q=x(1:num_q); qd=x(num_q+(1:num_q));
         h = obj.timestep;
+        z_inactive_guess_tol = .01;
```
Comment (avalenzu) [inline systems/plants/TimeSteppingRigidBodyManipulator.m]: Can we make this an settable  option? I imagine that it could have a performance impact depending on the speeds of the objects being simulated. Having dug through the Bullet source to modify tolerances like this, I'd hate to put our users through that.

File: systems/plants/TimeSteppingRigidBodyManipulator.m [hunk 2]
```diff
@@ -300,34 +279,104 @@ function checkDirty(obj)
...
+          end
+        else
+          phi_check = zeros(0,1);
+          J_check = zeros(0,num_q);
+          JL = zeros(0,num_q^2);
+        end
+        
         if (nContactPairs > 0)
           if (nargout>4)
             [phiC,normal,d,xA,xB,idxA,idxB,mu,n,D,dn,dD] = obj.manip.contactConstraints(q,true);
```
Comment (avalenzu) [inline systems/plants/TimeSteppingRigidBodyManipulator.m]: Should we make this `obj.manip.contactConstraints(q,false)` until we know that multiple contact points will work in this formulation? Right now we're relying on the fact that `collisionDetect` ignores this input argument.

File: systems/plants/TimeSteppingRigidBodyManipulator.m [hunk 3]
```diff
@@ -506,19 +549,18 @@ function checkDirty(obj)
...
-          z_inactive = z_inactive_guess;
+        z = zeros(nL+nP+(mC+2)*nC,1); 
+        if possible_indices_changed || isempty(obj.LCP_cache.data.z)
+          z_inactive = true(nL+nP+(mC+2)*nC,1);
           M_active = z_inactive;
+          obj.LCP_cache.data.z = z;
         else
-          z_inactive = obj.LCP_cache.data.z_inactive;
-          M_active = obj.LCP_cache.data.M_active;
+          z_inactive = obj.LCP_cache.data.z>lb+1e-8;
```
Comment (avalenzu) [inline systems/plants/TimeSteppingRigidBodyManipulator.m]: Maybe make this tolerance settable too? Or at least give it a name?

**Comment by avalenzu** [comment]
This looks good to me, though another pair of eyes on this might be good - so many indices to keep track of! The performance gains sound awesome.

File: systems/plants/ContactForceTorqueSensor.m [hunk 4]
```diff
@@ -119,7 +119,7 @@
       % contact normal and tangential directions in world coordinates
       nD = length(tsmanip.LCP_cache.data.contact_data.d);
       nC_body = nA + nB;
-      nC = manip.getNumContactPairs;
+      nC = size(tsmanip.LCP_cache.data.contact_data.normal,2);
```
Comment (mposa) [inline systems/plants/ContactForceTorqueSensor.m]: What's the reason behind this change? It's clear what getNumContactPairs does, and that it'll always return the same answer--the use of a cache is not as clear.

File: systems/plants/TimeSteppingRigidBodyManipulator.m (hunk 1 above)
Comment (mposa) [inline systems/plants/TimeSteppingRigidBodyManipulator.m]: Second this.

**Comment by mposa** [comment]
It looks good to me. Some of the cellfun usage was tough to parse (for the gradient calculations), so I'm mostly taking your word that it does the right thing. But the logic looks good to me.

File: systems/plants/ContactForceTorqueSensor.m (hunk 4 above)
Comment (RussTedrake) [inline systems/plants/ContactForceTorqueSensor.m]: because it will be a (changing) subset of all possible contact pairs now.  i've added a comment

File: systems/plants/TimeSteppingRigidBodyManipulator.m (hunk 2 above)
Comment (RussTedrake) [inline systems/plants/TimeSteppingRigidBodyManipulator.m]: i don't think we are relying on that.  multiple contact points support *should* work here once it's enabled.

File: systems/plants/TimeSteppingRigidBodyManipulator.m (hunk 1 above)
Comment (RussTedrake) [inline systems/plants/TimeSteppingRigidBodyManipulator.m]: done!
